### PR TITLE
NEW add OnlyTheseMembers Inherited Permission type

### LINKED ISSUE #10817: ONLY_THESE_USERS is poorly named and not as useful as it should be
## Affected Version

CMS all of them

## Description

The scope of `InheritedPermissions::ONLY_THESE_USERS` doesn't do what it says on the tin; instead, this functionality is better described as "only these GROUPS". Equally, this means you can't assign a specific user access to anything in this way - you have to jump through hoops to add them to a group first, which is often unnecessary double-handling.

## Proposal

**Minor release:** add a new permission  `InheritedPermissions::ONLY_THESE_INDIVIDUAL_USERS` that allows mapping to many `User`s, instead of groups. Mark this permission and the existing permission as deprecated. This adds the new functionality without breaking the existing API.

**Major release:** rename the above as follows:

`InheritedPermissions::ONLY_THESE_USERS` => `InheritedPermissions::ONLY_THESE_GROUPS`

`InheritedPermissions::ONLY_THESE_INDIVIDUAL_USERS` => `InheritedPermissions::ONLY_THESE_USERS`

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/10819
- https://github.com/silverstripe/silverstripe-assets/pull/556
- https://github.com/silverstripe/silverstripe-asset-admin/pull/1369
- https://github.com/silverstripe/silverstripe-cms/pull/2862
- https://github.com/silverstripe/silverstripe-admin/pull/1531
- https://github.com/silverstripe/developer-docs/pull/298

### ListboxField React PRs
- https://github.com/silverstripe/silverstripe-admin/pull/1533
- #10842

### PR DESCRIPTION
**Pull request by andrewandante**
To Do:

- [x] Update batch method SQL queries
- [x] Add tests
- [x] Make companion PR for silverstripe-assets
- [x] Make companion PR for silverstripe-asset-admin
- [x] Make companion PR for silverstripe-cms
- [x] Add proper deprecation

## Issue
- #10817

### CODE REVIEW

File: src/Security/InheritedPermissions.php
Comment (GuySartorelli) [inline src/Security/InheritedPermissions.php]: We can't add extra parameters (even if they have default values) in minor releases, since this means anyone crazy enough to subclass this thing and override this method would have to change the method signature in their class. It's technically a breaking change.

**Comment by andrewandante** [comment]
Test failures seem unrelated to this PR:

```
ERROR [Emergency]: Uncaught Symfony\Component\Translation\Exception\InvalidResourceException: Error parsing YAML, invalid file "/home/runner/work/silverstripe-framework/silverstripe-framework/vendor/silverstripe/versioned-admin/lang/en.yml". Message: Duplicate key "SilverStripe\Admin\Navigator\SilverStripeNavigatorItem" detected at line 36 (near "  WONT_BE_SHOWN: 'Note: this message will not be shown to your visitors'").
```

**Comment by GuySartorelli** [comment]
I've rerun the tests - all green

**Comment by andrewandante** [comment]
Thanks @GuySartorelli !

Should this be merged first so that the test suites in the other repos have something to latch on to? Or would you prefer to do it all in a group?

Also, am I pointing at the right branches etc? I'm never quite confident with that stuff 😅

**Comment by GuySartorelli** [comment]
All in a group, please. This one _will_ ultimately be merged first, but I'd like to make sure all of the PRs play nicely together first - and since they're all related it'll be easier to review them as a whole so the whole picture is in my mind when I do it.
Yup, those are the correct branches.

The only think that looks a little off at a quick glance is the todo (just remove that, and create a new CMS 6 issue for any changes that aren't behind a `@deprecated` phpdoc) and the deprecation message - the language for when there's a replacement is "Use [FQCN::method()] instead" (except where the new method is in the same class - then just omit the `FQCN::` bit) for both the phpdoc and the deprecation notice.

See https://docs.silverstripe.org/en/5/contributing/release_process/#deprecating-api and ignore the first part about `{@link <class>}` which needs to be reviewed (I know the `@link` style is nicer for IDEs but it's harder for us to work with when automating changelogs)

**Comment by andrewandante** [comment]
Righto - I've tested all the PRs and they are working as expected for me. I've pushed all the combined branches to a demo repo here: https://github.com/andrewandante/silverstripe-only-this-user-demo/blob/main/composer.json#L11-L23 if you feel like playing around.

Last remaining thing is that when applied to assets, the list of members renders as a CheckboxSetField, because there's no React form for the ListboxField. I don't know whether you consider this a blocker or not but I don't know that I can build that form with any confidence.

**Comment by GuySartorelli** [comment]
Thank you for this work, that's a great contribution 😁
I'm not sure when I'll get to reviewing this but know that it's on my radar. The `ListboxField` thing is a blocker, but I'll see if I can get a react component working for it, and if not I'll be able to request some specific changes on the PR to get it across the line.

**Comment by andrewandante** [comment]
Thanks! If I get a chance I'll take another swing at it - I got _reasonably_ close by copying the `TreeDropdownMultiSelect` stuff but then it needed, you know, a tree, and stopped working. The [ReactStrap Select Multiple](https://reactstrap.github.io/?path=/docs/components-forms--form) looks a bit gross, but could work in a pinch too - though it's not searchable.

**Comment by GuySartorelli** [comment]
It might help to look at [TagField](https://github.com/silverstripe/silverstripe-tagfield/blob/3/client/src/components/TagField.js) and [AnchorSelectField](https://github.com/silverstripe/silverstripe-cms/blob/5/client/src/components/AnchorSelectorField/AnchorSelectorField.js) which both use react-select (the same base library that `TreeDropdownField` uses)

**Comment by GuySartorelli** [comment]
Sorry to be a pain but can you please do the `ListboxField` enhancement as its own PR? It's a blocker for this work but it's really its own concern that should be reviewed and merged on its own merits (and as a bonus that might even mean someone else will get to reviewing it before I get around to looking at the rest of these)

**Comment by andrewandante** [comment]
> please do the ListboxField enhancement as its own PR

Yeah no worries, I'll split it out 👍

**Review by GuySartorelli — APPROVED** [review]
This all looks okay to me. Will merge once the other PRs have finished being reviewed and are approved.
After merging, I'll rerun CI on those other PRs before merging them as well.

**Review by GuySartorelli — APPROVED** [review]
Reapproving after the latest changes

**Pull request merged**
